docs/doxygen/man/man3/servers.3
.TH "servers" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
servers \- 
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBcreate_genesis_block\fP (const std::string genesis_file)"
.br
.ti -1c
.RI "void \fBinit_fast\fP (uint16_t node_count, hash_t pk)"
.br
.ti -1c
.RI "void \fBinit\fP (uint32_t newnow)"
.br
.ti -1c
.RI "void \fBwrite_start\fP ()"
.br
.ti -1c
.RI "uint32_t \fBread_start\fP ()"
.br
.ti -1c
.RI "void \fBxor4\fP (uint64_t *to, uint64_t *from)"
.br
.ti -1c
.RI "void \fBuser_csum\fP (\fBuser_t\fP &u, uint16_t \fBpeer\fP, uint32_t uid)"
.br
.ti -1c
.RI "void \fBinit_node_hash\fP (\fBnode\fP &nn)"
.br
.ti -1c
.RI "uint16_t \fBadd_node\fP (\fBuser_t\fP &ou)"
.br
.ti -1c
.RI "void \fBput_node\fP (\fBuser_t\fP &ou, uint16_t \fBpeer\fP)"
.br
.ti -1c
.RI "void \fBinit_user_hash\fP (\fBuser_t\fP &u)"
.br
.ti -1c
.RI "void \fBinit_user\fP (\fBuser_t\fP &u, uint16_t \fBpeer\fP, uint32_t uid, int64_t weight, uint8_t *pk, uint32_t when)"
.br
.ti -1c
.RI "void \fBput_user\fP (\fBuser_t\fP &u, uint16_t \fBpeer\fP, uint32_t uid)"
.br
.ti -1c
.RI "void \fBget_user\fP (\fBuser_t\fP &u, uint16_t \fBpeer\fP, uint32_t uid)"
.br
.ti -1c
.RI "bool \fBcheck_user\fP (uint16_t \fBpeer\fP, uint32_t uid)"
.br
.ti -1c
.RI "bool \fBfind_key\fP (uint8_t *pkey, uint8_t *skey)"
.br
.ti -1c
.RI "bool \fBfind_pkey\fP (uint8_t *pk)"
.br
.ti -1c
.RI "void \fBfind_more_keys\fP (uint8_t *pkey, std::map< uint16_t, \fBnodekey_t\fP > &nkeys)"
.br
.ti -1c
.RI "bool \fBcheck_nodehash\fP (uint16_t \fBpeer\fP)"
.br
.ti -1c
.RI "void \fBsave_undo\fP (uint16_t svid, std::map< uint32_t, \fBuser_t\fP > &undo, uint32_t users)"
.br
.ti -1c
.RI "void \fBclear_undo\fP ()"
.br
.ti -1c
.RI "void \fBget\fP ()"
.br
.ti -1c
.RI "void \fBget\fP (uint32_t path)"
.br
.ti -1c
.RI "void \fBread_servers\fP (uint32_t path, uint8_t *&data, uint32_t &len, uint8_t *&hash)"
.br
.ti -1c
.RI "void \fBput\fP ()"
.br
.ti -1c
.RI "bool \fBmsgl_put\fP (std::map< uint64_t, message_ptr > &map, char *cmphash)"
.br
.RI "\fIreplaced by \fBParser::MsglistParser\fP \fP"
.ti -1c
.RI "int \fBmsgl_get\fP (char *data)"
.br
.ti -1c
.RI "void \fBmsgl_map\fP (char *data, std::map< uint64_t, message_ptr > &map, uint16_t mysvid)"
.br
.ti -1c
.RI "int \fBmsgl_load\fP (std::map< uint64_t, message_ptr > &map, uint16_t mysvid)"
.br
.ti -1c
.RI "bool \fBmsgl_hash_tree_get\fP (uint16_t svid, uint32_t msid, uint32_t mnum, std::vector< \fBhash_s\fP > &hashes)"
.br
.RI "\fIreplaced by \fBParser::MsglistParser\fP \fP"
.ti -1c
.RI "uint16_t \fBget_vipuno\fP ()"
.br
.ti -1c
.RI "void \fBupdate_vipstatus\fP ()"
.br
.ti -1c
.RI "void \fBupdate_viphash\fP ()"
.br
.ti -1c
.RI "bool \fBvip_check\fP (uint8_t *viphash, uint8_t *vipkeys, int num)"
.br
.ti -1c
.RI "int \fBvip_size\fP (uint8_t *viphash)"
.br
.ti -1c
.RI "bool \fBload_vip\fP (int &len, char *&buf, uint8_t *vhash)"
.br
.ti -1c
.RI "void \fBfinish\fP ()"
.br
.ti -1c
.RI "void \fBhashnow\fP ()"
.br
.ti -1c
.RI "void \fBloadlink\fP (\fBheadlink_t\fP &link, uint32_t path, char *oldh)"
.br
.ti -1c
.RI "void \fBfilllink\fP (\fBheadlink_t\fP &link)"
.br
.ti -1c
.RI "void \fBloadhead\fP (\fBheader_t\fP &head)"
.br
.ti -1c
.RI "void \fBheader\fP (\fBheader_t\fP &head)"
.br
.ti -1c
.RI "int \fBdata_read\fP (const char *filename, bool read_nodes)"
.br
.ti -1c
.RI "int \fBdata_write\fP (const char *filename, bool write_nodes)"
.br
.ti -1c
.RI "int \fBheader_get\fP ()"
.br
.ti -1c
.RI "bool \fBsave_nowhash\fP (\fBheader_t\fP &head)"
.br
.ti -1c
.RI "bool \fBload_nowhash\fP ()"
.br
.ti -1c
.RI "void \fBheader_put\fP ()"
.br
.ti -1c
.RI "void \fBheader_last\fP ()"
.br
.ti -1c
.RI "void \fBheader_print\fP ()"
.br
.ti -1c
.RI "void \fBheader_print\fP (\fBheader_t\fP &head)"
.br
.ti -1c
.RI "void \fBsave_signature\fP (uint32_t path, uint16_t svid, uint8_t *sig, bool ok)"
.br
.ti -1c
.RI "bool \fBget_signatures\fP (uint32_t path, uint8_t *&data, uint32_t &nok)"
.br
.ti -1c
.RI "void \fBread_signatures\fP (uint32_t path, uint8_t *&data, uint32_t &nok, uint32_t &nno)"
.br
.ti -1c
.RI "bool \fBget_signatures\fP (uint32_t path, uint8_t *data, int nok, int nno)"
.br
.ti -1c
.RI "void \fBput_signatures\fP (\fBheader_t\fP &head, svsi_t *svsi)"
.br
.ti -1c
.RI "void \fBdel_signatures\fP ()"
.br
.ti -1c
.RI "void \fBcheck_signatures\fP (\fBheader_t\fP &head, svsi_t *svsi, bool save)"
.br
.ti -1c
.RI "bool \fBcopy_nodes\fP (\fBnode_t\fP *peer_node, uint16_t peer_srvn)"
.br
.ti -1c
.RI "int \fBcheck_nodes\fP (\fBnode_t\fP *peer_node, uint16_t peer_srvn, uint8_t *peer_nodehash)"
.br
.ti -1c
.RI "void \fBoverwrite\fP (\fBheader_t\fP &head, \fBnode_t\fP *peer_node)"
.br
.ti -1c
.RI "uint32_t \fBnextblock\fP ()"
.br
.ti -1c
.RI "void \fBblockdir\fP ()"
.br
.ti -1c
.RI "void \fBblockdir\fP (uint32_t blockTime)"
.br
.ti -1c
.RI "void \fBclean_old\fP (uint16_t svid)"
.br
.ti -1c
.RI "uint32_t \fBnowh32\fP (void)"
.br
.ti -1c
.RI "uint32_t \fBnowh64\fP (void)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "uint32_t \fBnow\fP"
.br
.ti -1c
.RI "uint32_t \fBmsg\fP"
.br
.ti -1c
.RI "uint32_t \fBnod\fP"
.br
.ti -1c
.RI "uint32_t \fBdiv\fP"
.br
.ti -1c
.RI "uint8_t \fBoldhash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint8_t \fBminhash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint8_t \fBmsghash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint8_t \fBnodhash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint8_t \fBviphash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint8_t \fBnowhash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "uint16_t \fBvok\fP"
.br
.ti -1c
.RI "uint16_t \fBvno\fP"
.br
.ti -1c
.RI "uint16_t \fBvtot\fP"
.br
.ti -1c
.RI "uint64_t \fBtxs\fP"
.br
.ti -1c
.RI "std::vector< \fBnode\fP > \fBnodes\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 83 of file servers\&.hpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/helper/servers.cpp
///TODO: uncomment logs when new log mechanism done
#include "servers.h"

#include <fstream>
#include <iostream>
#include <arpa/inet.h>
#include "default.hpp"
#include "hash.hpp"
#include "command/pods.h"
#include "parser/msglistparser.h"
#include "helper/ascii.h"
#include "helper/json.h"
#include "helper/blocks.h"
#include "helper/blockfilereader.h"

namespace Helper {

Servers::Servers() {
    m_filePath = "";
}

Servers::Servers(const char* filePath)
    : m_filePath(filePath) {
}

Servers::~Servers(){
}

void Servers::load(const char* filePath) {
    if (filePath) {
        m_filePath = filePath;
    }

    if (m_filePath.empty()) {
        return;
    }

    Helper::BlockFileReader file(m_filePath.c_str());
    if (file.isOpen()) {
        file.read((char*)&m_header, sizeof(m_header));
        for (unsigned int i=0; i<m_header.nodesCount; ++i) {
            ServersNode node;
            file.read((char*)&node, sizeof(node));
            m_nodes.push_back(node);
        }
    }
}

bool Servers::loadHeader() {
    if (m_filePath.empty()) {
        char filePath[64];
        if (!m_header.ttime) {
            snprintf(filePath, sizeof(filePath), "blk/header.hdr");
        } else {
            Helper::FileName::getName(filePath, m_header.ttime, "header.hdr");
        }
        m_filePath = filePath;
    }

    uint32_t check = m_header.ttime;

    Helper::BlockFileReader file(m_filePath.c_str());
    if (file.isOpen()) {
        file.read((char*)&m_header, sizeof(m_header));
        if (check && check != m_header.ttime) {
            return false;
        }
        return true;
    }
    return false;
}

bool Servers::loadNowhash() {
    if (m_filePath.empty()) {
        char filePath[64];
        sprintf(filePath, "blk/%03X.now", m_header.ttime>>20);
        m_filePath = filePath;
    }

    std::ifstream file(m_filePath, std::ifstream::in | std::ifstream::binary);
    if (file.is_open()) {
        file.seekg(((m_header.ttime&0xFFFFF)/BLOCKSEC)*SHA256_DIGEST_LENGTH);
        file.read((char*)&m_header.nowHash, sizeof(m_header.nowHash));
        if (!file) {
            return false;
        }
        file.close();
    } else {
        return false;
    }
    return true;
}

void Servers::setNow(uint32_t time) {
    m_header.ttime = time;
}

uint8_t* Servers::getNowHash() {
    return m_header.nowHash;
}

ServersHeader Servers::getHeader() {
    return m_header;
}

std::vector<ServersNode> Servers::getNodes() {
    return m_nodes;
}

ServersNode Servers::getNode(unsigned int nodeId) {
    if (m_nodes.size() < nodeId){
        throw std::invalid_argument("Node Id over limit");
    }

    return m_nodes.at(nodeId);
}

unsigned int Servers::getNodesCount() {
    return m_nodes.size();
}

bool Servers::getMsglHashTree(uint16_t svid,uint32_t msid,uint32_t msg_number,std::vector<hash_s>& hashes) {
    if (!msg_number) {
        return false;
    }

    if (!loadHeader() && (m_header.messageCount == 0)) {
        return false;
    }

    Parser::MsglistParser parser(m_header.ttime);
    if (!parser.load() || parser.isEmpty()) {
//        DLOG("ERROR %s not found\n",filename);
        return false;
    }

    uint16_t msg_svid = 0;
    uint32_t msg_msid = 0;
    hash_s msg_hash;
    Parser::MsgIterator it = parser.begin() + (msg_number - 1);
    msg_svid = it->node_id;
    msg_msid = it->node_msid;

    if (msg_svid != svid || msg_msid != msid) {
//        DLOG("ERROR bad index %d %04X:%08X <> %04X:%08X\n", mnum, svid,msid, msg_svid, msg_msid);
        return false;
    }

    if (msg_number%2 == 0) {
        --it;
        memcpy(&msg_hash, it->hash, sizeof(msg_hash));
        hashes.push_back(msg_hash);
    } else {
        ++it;
        if (it != parser.end()) {
            memcpy(&msg_hash, it->hash, sizeof(msg_hash));
            hashes.push_back(msg_hash);
        }
    }

    --msg_number;
    uint32_t htot = parser.getHashesTotalSize();
    hashtree tree;
    if (htot > 0) {
        std::vector<uint32_t> add;
        tree.hashpath((msg_number)/2, (m_header.messageCount+1)/2, add);
        Parser::HashIterator hIt;
        for (auto n : add) {
            hIt = parser.h_begin() + n;
            if (hIt == parser.h_end()) {
                it = parser.end() - 1;
                hash_s lastTxnHash;
                std::copy(it->hash, it->hash+32, lastTxnHash.hash);
                hashes.push_back(lastTxnHash);
            } else {
                hashes.push_back(*hIt);
            }
        }
    }

    //DEBUG confirm hash
    hash_t nhash;
    tree.hashpathrun(nhash,hashes);
    if(memcmp(m_header.msgHash, nhash,32)) {
//        DLOG("HASHTREE failed (path len:%d) to get msghash\n",(int)hashes.size());
        return false;
    }
    hash_s* hashmin=(hash_s*)m_header.minHash;
    hashes.push_back(*hashmin);
    tree.addhash(nhash, m_header.minHash);
    if(memcmp(m_header.nowHash, nhash, 32)) {
//        DLOG("HASHTREE failed (path len:%d) to get nowhash\n",(int)hashes.size());
        return false;
    }

    return true;
}

void ServersNode::toJson(boost::property_tree::ptree& node)
{
    char value[65];
    value[64]='\0';

    Helper::ed25519_key2text(value, publicKey, 32);
    node.put("public_key", value);
    Helper::ed25519_key2text(value, (uint8_t*)hash, 32);
    node.put("hash", value);
    Helper::ed25519_key2text(value, messageHash, 32);
    node.put("message_hash", value);
    node.put("msid", messageId);
    node.put("mtim", messageTime);
    node.put("balance", print_amount(weight));
    node.put("status", status);
    node.put("account_count", accountCount);
    node.put("port", port);

    struct in_addr ip_addr;
    ip_addr.s_addr = ipv4;
    node.put("ipv4", inet_ntoa(ip_addr));
}

}
